addr_manager_badger.go

```diff
@@ -0,0 +1,247 @@
+package peerstore
+
+import (
+	"context"
+	"time"
+
+	"github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/dgraph-io/badger"
+	"github.com/multiformats/go-multihash"
+	"encoding/gob"
+	"bytes"
+)
+
+type BadgerAddrManager struct {
+	DB *badger.DB
+	addrSubs map[peer.ID][]*addrSub
+}
+
+type addrentry struct {
+	Addr []byte
+	TTL time.Duration
+}
+
+func (mgr *BadgerAddrManager) sendSubscriptionUpdates(p *peer.ID, addrs []ma.Multiaddr) {
+	subs := mgr.addrSubs[*p]
+	for _, sub := range subs {
+		for _, addr := range addrs {
+			sub.pubAddr(addr)
+		}
+	}
+}
+
+func (mgr *BadgerAddrManager) Close() {
+	if err := mgr.DB.Close(); err != nil {
+		log.Error(err)
+	}
+}
+
+func NewBadgerAddrManager(dataPath string) (*BadgerAddrManager, error) {
+	opts := badger.DefaultOptions
+	opts.Dir = dataPath
+	opts.ValueDir = dataPath
+	db, err := badger.Open(opts)
+	if err != nil {
+		return nil, err
+	}
+	return &BadgerAddrManager{DB: db}, nil
+}
+
+func (mgr *BadgerAddrManager) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+// use murmur3 because it's the most compact
+func hashMultiaddr(addr *ma.Multiaddr) ([]byte, error) {
+	return multihash.Encode((*addr).Bytes(), multihash.MURMUR3)
+}
+
+func createUniqueKey(p *peer.ID, addr *ma.Multiaddr) ([]byte, error) {
+	prefix := []byte(*p)
+	addrHash, err := hashMultiaddr(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return append(prefix, addrHash...), nil
+}
+
+func addAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, txn *badger.Txn) {
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		entry := &addrentry{Addr: addr.Bytes(), TTL: ttl}
+		key, err := createUniqueKey(&p, &addr)
+		if err != nil {
+			log.Error(err)
+			txn.Discard()
+			return
+		}
+		buf := &bytes.Buffer{}
+		enc := gob.NewEncoder(buf)
+		if err := enc.Encode(entry); err != nil {
+			log.Error(err)
+			txn.Discard()
+			return
+		}
+		txn.SetWithTTL(key, buf.Bytes(), ttl)
+	}
+}
+
+func (mgr *BadgerAddrManager) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	if ttl <= 0 {
+		log.Debugf("short circuiting AddAddrs with ttl %d", ttl)
+		return
+	}
+
+	txn := mgr.DB.NewTransaction(true)
+	defer txn.Discard()
+
+
+	go mgr.sendSubscriptionUpdates(&p, addrs)
+	addAddrs(p, addrs, ttl, txn)
+
+	txn.Commit(nil)
+}
+
+func (mgr *BadgerAddrManager) SetAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
+	mgr.SetAddrs(p, []ma.Multiaddr{addr}, ttl)
+}
+
+func (mgr *BadgerAddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
+	txn := mgr.DB.NewTransaction(true)
+	defer txn.Discard()
+
+	for _, addr := range addrs {
+		if addr == nil {
+			continue
+		}
+		key, err := createUniqueKey(&p, &addr)
+		if err != nil {
+			log.Error(err)
+			continue
+		}
+		if ttl <= 0 {
+			if err := txn.Delete(key); err != nil {
+				log.Error(err)
+			}
```

Comment:
If you add a `continue` here, the following `else` clause is no longer required.